Fix Issue 19415 - return non-copyable struct fails if member function has return attribute

RazvanN7 opened the pull request:
```d
struct S
{
   int x;
   S foo() return { return S(x); }
   this(this) @disable;
}

S bar()
{
   S s;
   return s; // Error: struct `S` is not copyable because it is annotated with @disable
}
```

When `struct S` is semantically analyzed, the field StructDeclaration.postblit is null. Before actually building the postblit and setting StructDeclaration.postblit to point to it, all the members of `struct S` are semantically analyzed. This results in the fact that at the time when foo is analyzed, the struct does not appear to have a postblit. During the semantic analysis of foo, this line [1] is executed in order to delete the `return` attribute from the function. The code path is : TypeStruct.hasPointers() -> AggregateDeclaration.size() -> AggregateDeclaration.determineSize() -> StructDeclaration.finalizeSize(). At the end of finalizeSize this is done: [2]. The code path is: toArgTypes -> StructDeclaration.isPOD. Although the struct is not pod because it has a postblit, because the struct has not been semantically analyzed completely, it appears as it is POD, messing with the size of the struct and ultimately making it wrongfully not usable with NRVO.

The fix makes it so that AggregateDeclaration.size() is not called anymore to verify forward references. This should also improve performance since a lot of unnecessary checks are avoided.

This has truly been an adventure.

[1] https://github.com/dlang/dmd/pull/8346/files#diff-64d4abe08ba2355c2b95cb22e46b8572R876
[2] https://github.com/dlang/dmd/blob/master/src/dmd/dstruct.d#L429

RazvanN7 committed:
Fix Issue 19415 - return non-copyable struct fails if member function has return attribute

schveiguy commented:
Seems to fail for 64-bit builds, but when I look at the log, I can't figure out why. Looks like linker errors, but no actual message. Seems like it's `std.container.util` unittests.

schveiguy commented:
You have the issue number messed up, should be 19415, not 19145.

jacob-carlborg commented:
I would be nice to have the PR description i the commit message. Doing that from the beginning would automatically add the commit message as the PR description.

RazvanN7 commented:
> Seems to fail for 64-bit builds, but when I look at the log, I can't figure out why. Looks like linker errors, but no actual message. Seems like it's std.container.util unittests.

I found a better fix. It seems like StructDeclaration.size() was used abusively to check for forward references. I added the single check; this is correct and even more efficient.

> You have the issue number messed up, should be 19415, not 19145.

Fixed.

 > I would be nice to have the PR description i the commit message. Doing that from the beginning would automatically add the commit message as the PR description.

I think it would be a bit too verbose.
